### PR TITLE
[Connected] Preventing loss of TextField focus caused by shifting DOM position

### PR DESCRIPTION
**Pull request by arapehl**
### WHY are these changes introduced?

Fixes: https://github.com/Shopify/polaris-react/issues/2169
Fixes: https://github.com/Shopify/online-store-web/issues/4181

### WHAT is this pull request doing?

Ensuring that the `Connected` component always returns fragments at the same DOM depth so as not to cause a `TextField`'s `input` field to unmount and lose focus.

<details>
  <summary>Introducing a "Search" button (`connectedRight`) while typing causes a loss of focus</summary>
  
</details>

<details>
  <summary>Ensuring maintained focus by always returning a TextField at the same depth in the DOM (what this PR does)</summary>
  
</details>

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}

```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

### CODE REVIEW

File: src/components/TextField/tests/TextField.test.tsx [hunk 1]
```diff
@@ -1042,15 +1039,13 @@ describe('<TextField />', () => {
...
 
       connectedInteriorWrapper.simulate('focus');
 
       connectedInteriorWrapper = textField
         .find(Connected)
-        .find('div')
-        .first();
+        .find('div.TextField');
 
       expect(connectedInteriorWrapper.hasClass('focus')).toBe(true);
```
Comment (dleroux) [inline src/components/TextField/tests/TextField.test.tsx]: Same as in web we follow their best practice and avoid using classNames in tests. [see here](https://github.com/Shopify/web-foundation/blob/master/Best%20practices/React/Testing.md#implementation-details). `document.activeElement` would be more reliable here.

Comment (arapehl) [reply src/components/TextField/tests/TextField.test.tsx]: The test is called `applies focus variant style onFocus`. Its whole purpose is to test for the `focus` class name. I was just getting it to work given the change in DOM output caused by this PR. The previous test called `sets focus to the <input /> onClick` checks for `document.activeElement` albeit from a different trigger.

I can ditch the whole test, if you'd like.

Comment (danrosenthal) [reply src/components/TextField/tests/TextField.test.tsx]: Dan's guidance on asserting `document.activeElement` is probably a better approach here.

File: src/components/TextField/tests/TextField.test.tsx [hunk 2]
```diff
@@ -1040,19 +1040,18 @@ describe('<TextField />', () => {
       const textField = mountWithAppProvider(
```
Comment (asalisbury) [inline src/components/TextField/tests/TextField.test.tsx]: What is a "focus variant style"?

File: src/components/TextField/tests/TextField.test.tsx [hunk 3]
```diff
@@ -1040,19 +1040,18 @@ describe('<TextField />', () => {
...
-      connectedInteriorWrapper = textField
+      const connectedChild = textField
         .find(Connected)
-        .find('div')
-        .first();
+        .prop('children') as ReactElement;
 
-      expect(connectedInteriorWrapper.hasClass('focus')).toBe(true);
+      act(() => {
+        connectedChild.props.onFocus({});
```
Comment (asalisbury) [inline src/components/TextField/tests/TextField.test.tsx]: What does this do? I think the assertion passes even if you 🔥 it.

File: src/components/TextField/tests/TextField.test.tsx [hunk 4]
```diff
@@ -1026,10 +1027,9 @@ describe('<TextField />', () => {
       );
       const connectedChild = textField
         .find(Connected)
-        .find('div')
-        .first();
+        .prop('children') as ReactElement;
```
Comment (asalisbury) [inline src/components/TextField/tests/TextField.test.tsx]: Could we add the inverse assertion above the click? 

```tsx
expect(textField.getDOMNode().querySelector('input')).not.toStrictEqual(
  document.activeElement!,
);
```

**Comment by arapehl** [comment]
So, I ended up removing the `onFocus` test because it doesn't actually test anything apart from setting a `focus` class name. I also added an additional assertion to the `onClick` test to makes sure that it wasn't originally the active element and that the `onClick` caused it to be.

**Comment by asalisbury** [comment]
:shipit:

**Pull request merged**